Update EIP-7761: Change HASCODE to EXTCODETYPE (#9229)
ethereum/EIPs: EIPS/eip-7761.md

## Abstract

Allow EOF contracts to discriminate between EOAs (Externally Owned Accounts) and contract accounts by introducing an `EXTCODETYPE` instruction.

## Motivation

EOFv1 as scoped in [EIP-7692] removes code introspection capabilities from the EVM, including the `EXTCODESIZE` instruction (in [EIP-3540]). This makes it hard for [ERC-721] and [ERC-1155] standard contracts to be implemented, as they rely on discovering whether a token's `safeTransfer` call target was an EOA or a contract account:

- `safeTransfers` to EOAs succeed
- `safeTransfers` to contract accounts call an `onERC721Received` (`onERC1155Received`) on them and expect to get a special magic return value, otherwise the transfer reverts (on the assumption that such a recipient may not be able to interact with the token)

> Appliction and library developers are also concerned about dynamic proxies written in EOF accidentally pointing to legacy accounts which cannot safely be called with EXTDELEGATECALL. They would like a way to differentiate between EOF and legacy contracts.

Reviewer: ```suggestion
Application and library developers are also concerned about dynamic proxies written in EOF accidentally pointing to legacy accounts which cannot safely be called with EXTDELEGATECALL. They would like a way to differentiate between EOF and legacy contracts.
```

> Application and library developers are also concerned about dynamic proxies written in EOF accidentally pointing to legacy accounts which cannot safely be called with EXTDELEGATECALL. They would like a way to differentiate between EOF and legacy contracts.

Reviewer 2: ```suggestion
Application and library developers are also concerned about dynamic proxies written in EOF accidentally pointing to legacy accounts which cannot safely be called with `EXTDELEGATECALL`. They would like a way to differentiate between EOF and legacy contracts.
```

`EXTCODETYPE` is aimed to fill this gap and bring back the possibility to easily implement ERC-721 and ERC-1155 standard contracts in EOF as well as preserve EOF proxy contract safety.

## Specification

### Parameters

| Constant                  | Value                                                              |
|---------------------------|--------------------------------------------------------------------|
| `FORK_BLKNUM`             | tbd                                                                |
| `GAS_COLD_ACCOUNT_ACCESS` | Defined as `2600` in the [Ethereum Execution Layer Spec Constants] |
| `GAS_WARM_ACCESS`         | Defined as `100` in the [Ethereum Execution Layer Spec Constants]  |
| `TYPE_EOA`                | 0                                                                  |
| `TYPE_LEGACY_CONTRACT`    | 1                                                                  |

> | `TYPE_EOF_CONTRACT`       | 2                                                                  |

Reviewer: The spec below uses `ACCOUNT_TYPE_*` constants.

We introduce a new EOFv1 instruction on the block number `FORK_BLKNUM`: `EXTCODETYPE` (`0xe9`)

EOF code which contains this instruction before `FORK_BLKNUM` is considered invalid. Beginning with block `FORK_BLKNUM` `0xe9` is added to the set of valid EOFv1 instructions.

### Execution Semantics
    
#### `EXTCODETYPE`

- deduct `GAS_WARM_ACCESS` gas
- pop 1 argument `target_address` from the stack
- if `target_address` has any of the high 12 bytes set to a non-zero value (i.e. it does not contain a 20-byte address), then halt with an exceptional failure
  - Notice: Future expansion of the EVM address space may remove or reduce the number of bytes in this check.
- deduct `GAS_COLD_ACCOUNT_ACCESS - GAS_WARM_ACCESS` if `target_address` is not in `accessed_addresses` and add `target_address` to `accessed_addresses`
- Load the code from `target_address` and refer to it as `loaded_code` 

> - if `loaded_code` is empty (does not exist or has zero length), push `ACCOUNT_TYPE_EOA` onto the stack and stop processing the operation

Author: Hmm... yes.  Because an account can gain code.

It's not an empty account, those are accounts without balance, code, or nonce.  It may have balance.

I'm thinking `TYPE_NONE` - code type is it has no code.

This also make the name `EXTCODETYPE` better than `ACCOUNTTYPE`, because accounts shouldn't change type?

> - if `loaded_code` is empty (does not exist or has zero length), push `ACCOUNT_TYPE_EOA` onto the stack and stop processing the operation

Reviewer: `TYPE_NONE` is good! And yes I think `EXTCODETYPE` is a good name.

2. Continuing to use legacy EVM themselves. This is suboptimal, since EVM compilers are likely to at some point deprecate legacy EVM as compilation target
3. Updating the standards to not rely on code introspection patterns in `safeTransfer` safeguards. This can be accomplished for example by leveraging [ERC-165], leaving out only contracts which do not implement ERC-165, and at the same time have non-throwing fallback functions, as indistinguishable from EOAs. This is not easy to achieve since ERC-721 and ERC-1155 are Final and adopted in practice.

### "Endgame Account Abstraction" issues

`EXTCODETYPE` (and earlier `EXTCODESIZE` available in legacy EVM) are claimed to slow down AA adoption, because they encourage patterns which discriminate between smart contract and EOA accounts, e.g. not allowing the former to interact. However, there are counterarguments that it is up to other factors which slow down AA adoption (assumption that accounts can produce ECDSA signatures, and the lack of adoption of smart contract signatures).

### Including safeguarding against proxy bricking
    

> In parallel to the ERC-721 / ERC-1155 problem, another potential risk has been brought to attention. Since EOFv1 prohibits `EXTDELEGATECALL` targeting legacy contracts, there exists a scenario where an EOF proxy contract accidentally upgrades its implementation to a legacy EVM one. Since reverting this or upgrading again (using current proxy standards) requires the implementation contract to be called, it would effectively render the contract unusable.

Author: It is in the motivation, line 25, jus not with the full technical drill down.

> For this reason it was decided to have a generalized `EXTCODETYPE` instruction instead of a `HASCODE` instruction. This providing toe means to safeguard against such a scenario that a simpler instruction could not.

Reviewer 2: ```suggestion
For this reason it was decided to have a generalized `EXTCODETYPE` instruction instead of a `HASCODE` instruction. This provides the means to safeguard against such a scenario that a simpler instruction could not.
```

### Relation to [EIP-7702] "Set EOA account code"

> After [EIP-7702] is activated, the discrimination between EOAs and contract accounts using `EXTCODESIZE` (or `ACCOUNT`) has an edge case: Whenever an EOA sets its code to a contract account which does not respond as expected to an `onERC721Received` (`onERC1155Received`) callback, transfers to it will revert, despite the recipient being able to interact with the token. This has been deemed unlikely to be a problem, as for the intended real-world uses of EIP-7702, those callbacks will be implemented by designator codes.

Reviewer 2: ```suggestion
After [EIP-7702] is activated, the discrimination between EOAs and contract accounts using `EXTCODESIZE` (or `EXTCODETYPE`) has an edge case: Whenever an EOA sets its code to a contract account which does not respond as expected to an `onERC721Received` (`onERC1155Received`) callback, transfers to it will revert, despite the recipient being able to interact with the token. This has been deemed unlikely to be a problem, as for the intended real-world uses of EIP-7702, those callbacks will be implemented by designator codes.
```

Because of the requirement that EOF proxy contracts be able to determine if the new target is safe we cannot return a value for indicating that the account itself is delegated without also providing the delegated address. Instead, we return  the code for the account that has the delegated code, applying the delegation. This is because the executable code is what matters for proxy updates.

> This differs from EXTCODECOPY and EXTCODEHASH behavior which do return a hash and short code indicating the proxy.

Reviewer: A suggestion:

```suggestion
This differs from EXTCODECOPY and EXTCODEHASH behavior which do return a hash and short code indicating the proxy. While EXTCODETYPE is a "code reading operation" its purpose is to describe the execution behavior of the account.
```

This references this part of the EIP-7702 spec:
> This designator requires all code executing operations to follow the address pointer to get the account’s executable code, and **requires all other code reading operations to act only on the first 2 bytes of the designator** (0xef01).

> This differs from EXTCODECOPY and EXTCODEHASH behavior which do return a hash and short code indicating the proxy. The behavior of EXTCODETYPE is closer to a "code reading operation" in purpose as it is meant to describe the execution behavior of the account.

Reviewer 2: I think you meant "executing" here? Executing operations follow delegation (as EXTCODETYPE), reading operations read the delegation designator.
```suggestion
This differs from `EXTCODECOPY` and `EXTCODEHASH` behavior which do return a hash and short code indicating the proxy. The behavior of `EXTCODETYPE` is closer to a "code executing operation" in purpose as it is meant to describe the execution behavior of the account.
```
